README.md
[![Build Status](https://travis-ci.org/sixapart/Router-Assemble.svg?branch=master)](https://travis-ci.org/sixapart/Router-Assemble) [![Coverage Status](https://coveralls.io/repos/github/sixapart/Router-Assemble/badge.svg?branch=master)](https://coveralls.io/github/sixapart/Router-Assemble?branch=master)

# NAME

Router::Assemble - Fast, dumpable/restorable and weight adjustable routing engine for web applications

# SYNOPSIS

    use Router::Assemble;

    my $router = Router::Assemble->new();
    $router->add('/', 'dispatch_root');
    $router->add('/entrylist', 'dispatch_entrylist');
    $router->add('/:user', 'dispatch_user');
    $router->add('/:user/{year}', 'dispatch_year');
    $router->add('/:user/{year}/{month:\d+}', 'dispatch_month');
    $router->add('/download/*', 'dispatch_download');

    my $dest = $router->match($env->{PATH_INFO});

# DESCRIPTION

Router::Assemble is yet another routing engine for web applications. This module is inspired by [Router::Boom](https://metacpan.org/pod/Router::Boom) and powered by [Regexp::Assemble](https://metacpan.org/pod/Regexp::Assemble).

## Fast

This module is pretty fast, almost as fast as [Router::Boom](https://metacpan.org/pod/Router::Boom).

                         Rate Router::Assemble     Router::Boom
    Router::Assemble 220553/s               --              -0%
    Router::Boom     220554/s               0%               --

## Dump and restore

    my $router = Router::Assemble->new();
    build_routes($router); # add routes for your application

    # Get an instance of Router::Assemble::Matcher.
    # This object holds the minimum information necessary for routing.
    my $matcher = $router->matcher;
    $matcher->match($env->{PATH_INFO});

    # dump and restore
    my $hash = $matcher->to_hash;                     # Dump as a hash object that can be expressed as JSON
    my $m    = Router::Assemble::Matcher->new($hash); # Restore. This is very fast
    $m->match($env->{PATH_INFO});

## Weight adjustable

By default, [Router::Assemble](https://metacpan.org/pod/Router::Assemble) routes in order of addition.

    my $router = Router::Assemble->new();
    $router->add('/{category:\w+}', 'cat');
    $router->add('/blog',           'blog_main');

    $router->match('/blog') # cat

By using _route\_cmp_, we can adjust routing order.

    my $router = Router::Assemble->new(
        route_cmp => sub($$){
            my ($a, $b) = @_;
            # route to static route first
            $a->[0] =~ m/[\+\*]/ <=> $b->[0] =~ m/[\+\*]/;
        },
    );
    $router->add('/{category:\w+}', 'cat');
    $router->add('/blog',           'blog_main');

    $router->match('/blog') # blog_main

## Compatibility with [Router::Boom](https://metacpan.org/pod/Router::Boom)

[Router::Assemble](https://metacpan.org/pod/Router::Assemble) is compatible with [Router::Boom](https://metacpan.org/pod/Router::Boom) about basic API.

# MEHTODS

- my $router = Router::Assemble->new(%options)

    Create a new instance. %options may contains these values.

    - route\_cmp

        A subroutine for comparing routes. this subroutine will be called at first arguments of "sort". The value to be compared is \[$path, $context\] pair of registered via "add".

- $router->add($path, $context)

    Add a route. $path accepts a string value. See the PATH DEFINITION section for details on how $path value is specified. $context accepts any value.

- my ($context, $captured) = $router->match($path);

    Attempt to find route that matches the given $path and return two values if a route is found. $context is a value of registered via "add". $captured is a hashref that contains placeholders given from $path.

    Return empty list if no route is found.

# PATH DEFINITION

- `{name:\w+}`

    Accepts "\\w+" in this component. And "name" will be used as a key in the $captured.

        $router->add('/user/{name:\w+}', 'profile');
        $router->match('/user/foo');
        # => ('profile', {name => 'foo'})

    You can use "(?:pattern)". (But you can not use "(pattern)")

        $router->add('/user/{name:(?:bar|baz)}', 'special_profile');
        $router->add('/user/{name:\w+}', 'profile');
        $router->match('/user/bar');
        # => ('special_profile', {name => 'bar'})

- `:name` and `{name}`

    Same as `{name:[^/]+}`.

        $router->add('/user/:name', 'profile');
        $router->match('/user/foo');
        # => ('profile', {name => 'foo'})

- `*`

    Same as `{*:.+}`.

        $router->add('/archive/*', 'archive');
        $router->match('/archive/path/to/files.zip');
        # => ('archive', {'*' => 'path/to/files.zip'})

# LICENSE

Copyright (C) Six Apart Ltd. <sixapart@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Taku AMANO <usualoma@cpan.org>

# SEE ALSO

[Router::Boom](https://metacpan.org/pod/Router::Boom) Pretty fast routing engine for web applications.

[Regexp::Assemble](https://metacpan.org/pod/Regexp::Assemble) Assemble multiple Regular Expressions into a single RE.


Changes
Revision history for Perl extension Router-Assemble

{{$NEXT}}
    - original version


author/benchmark.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;
use 5.010000;
use autodie;
use Benchmark qw(:all);
use Router::Boom;
use Router::Assemble;

my $router_boom = do {
    my $router = Router::Boom->new();
    $router->add('/',      'Root');
    $router->add('/entrylist', 'EntryList');
    $router->add("/$_", "$_") for 'a'..'z';
    $router->add('/:user', 'User#index');
    $router->add('/:user/:year/', 'UserBlog#year_archive');
    $router->add('/:user/:year/:month/', 'UserBlog#month_archive');
    $router;
};

my $router_assemble = do {
    my $router = Router::Assemble->new();
    $router->add('/',      'Root');
    $router->add('/entrylist', 'EntryList');
    $router->add("/$_", "$_") for 'a'..'z';
    $router->add('/:user', 'User#index');
    $router->add('/:user/:year/', 'UserBlog#year_archive');
    $router->add('/:user/:year/:month/', 'UserBlog#month_archive');
    $router->matcher;
};

cmpthese(
    -1, {
        'Router::Boom'     => sub { $router_boom->match('/dankogai/2013/02/') },
        'Router::Assemble' => sub { $router_boom->match('/dankogai/2013/02/') },
    }
);



lib/Router/Assemble/Util.pm
package Router::Assemble::Util;

use strict;
use warnings;
use utf8;

use parent 'Exporter';
our @EXPORT = qw(
    normalize_path_component split_path is_placeholder split_placeholder
);

our $placeholder_content_regexp = qr/(?:\{[0-9,]+\}|[^{}])/;

sub normalize_path_component {
    my $c = shift;

    if (   $c =~ m/\A\{((?:$placeholder_content_regexp(?!:))+)\}\z/
        || $c =~ m/\A:([A-Za-z0-9_]+)\z/ )
    {
        "{$1:[^/]+}";
    }
    elsif ( $c eq '*' ) {
        '{*:.+}';
    }
    else {
        $c;
    }
}

sub split_path {
    $_[0] =~ m{(
        \{(?:$placeholder_content_regexp)+\} | # /blog/{year:\d{4}}
        :[A-Za-z0-9_]+                       | # /blog/:year
        \*                                   | # /blog/*/*
        (?:\(\?\:.*?\)|[^\{:*])+               # normal string
    )}gx
}

sub is_placeholder {
    !!split_placeholder( $_[0] );
}

sub split_placeholder {
    $_[0] =~ m{\A\{
        ((?:$placeholder_content_regexp)+?)
        :
        ((?:$placeholder_content_regexp)+)
    \}\z}x;
}

1;


lib/Router/Assemble/Builder.pm
package Router::Assemble::Builder;

use strict;
use warnings;
use utf8;

use Carp ();

use Router::Assemble::Assembler;
use Router::Assemble::Matcher;
use Router::Assemble::Util;
use Router::Assemble::Constant;

sub new {
    my $class = shift;
    my $self  = +{@_==1 ? %{$_[0]} : @_};

    $self->{routes}                = +[] unless exists $self->{routes};
    $self->{quotemeta_static_path} = 1   unless exists $self->{quotemeta_static_path};

    bless $self, $class;
}

sub new_assembler {
    Router::Assemble::Assembler->new;
}

sub compile {
    use re 'eval';

    my $self = shift;
    my $assembler = $self->new_assembler;

    my @routes = @{$self->{routes}};
    if (my $cmp = $self->{route_cmp}) {
        @routes = sort $cmp @routes;
    }

    my @compiled_routes;
    for (my $i = 0; $i < scalar @routes; $i++) {
        my @names;
        my @tokens =
            map {
                if (is_placeholder($_)) {
                    my ($name, $regexp) = split_placeholder($_);
                    push @names, $name;
                    "($regexp)";
                }
                else {
                    $self->{quotemeta_static_path} ? quotemeta($_) : $_;
                }
            }
            map { normalize_path_component($_) }
            split_path($routes[$i][0]);

        $assembler->add(join('', @tokens) . "\\z(?<@{[INDEX_KEYWORD]}>$i)");
        push @compiled_routes, [\@names, $routes[$i][1]];
    }

    (my $r = "\\A@{[$assembler->regexp]}") =~ s/\(\?<@{[INDEX_KEYWORD]}>(\d+)\)/(?{$1})/g;

    [qr/$r/, \@compiled_routes];
}

# True if : ()
# False if : (?:)
sub _is_normal_capture {
    $_[0] =~ /
        \(
            (?!
                \?:
            )
    /x
}

sub add {
    my $self = shift;
    my ($path, $ctx) = @_;

    delete $self->{matcher};

    Carp::croak("You can't include parens in your custom rule.")
        if grep {
                _is_normal_capture(
                    is_placeholder($_)
                        ? (split_placeholder($_))[1]
                        : $_
                )
            }
            grep { !$self->{quotemeta_static_path} || is_placeholder($_) }
            map { normalize_path_component($_) }
            split_path($path);

    for (my $i = 0; $i < scalar @{$self->{routes}}; $i++) {
        next if $self->{routes}->[$i][0] ne $path;
        $self->{routes}->[$i][1] = $ctx;
        return;
    }

    push @{$self->{routes}}, [$path, $ctx];
}

sub matcher {
    my $self = shift;

    $self->{matcher} ||= do {
        my ($regexp, $routes) = @{$self->compile};
        Router::Assemble::Matcher->new(
            regexp => $regexp,
            routes => $routes,
        );
    };
}

1;


lib/Router/Assemble/Matcher.pm
package Router::Assemble::Matcher;

use strict;
use warnings;
use utf8;

use Router::Assemble::Util;

sub new {
    use re 'eval';

    my $class = shift;
    my $self  = +{@_==1 ? %{$_[0]} : @_};

    $self->{routes} = +[] unless exists $self->{routes};
    $self->{regexp} = qr/$self->{regexp}/ if exists $self->{regexp};

    bless $self, $class;
}

sub regexp {
    my $self = shift;
    $self->{regexp};
}

sub match {
    my $self = shift;
    my ($path) = @_;

    # copatibility with Router::Simple/Router::Boom
    #
    # "I think there was a discussion about that a while ago and it is up to apps to deal with empty PATH_INFO as root / iirc"
    # -- by @miyagawa
    #
    # see http://blog.64p.org/entry/2012/10/05/132354
    $path = '/' if $path eq '';

    my @res = ($path =~ $self->{regexp})
        or return ();
    my ($names, $ctx) = @{$self->{routes}->[$^R]};

    my %captured;
    @captured{@$names} = grep { defined $_ } @res;

    $ctx, \%captured;
}

sub to_hash {
    my $self = shift;

    +{
        regexp => $self->{regexp} . "",
        routes => $self->{routes},
    };
}

1;


lib/Router/Assemble/Assembler.pm
package Router::Assemble::Assembler;

use strict;
use warnings;
use utf8;

use List::Util qw(min);
use Regexp::Assemble;

use Router::Assemble::Constant;

sub new {
    my $class = shift;
    my $self = +{@_==1 ? %{$_[0]} : @_};

    local $Regexp::Assemble::Current_Lexer = qr/(?![[(\\]).(?:[*+?]\??|\{\d+(?:,\d*)?\}\??)?|\\(?:[bABCEGLQUXZ]|[lu].|(?:[^\w]|[aefnrtdDwWsS]|c.|0\d{2}|x(?:[\da-fA-F]{2}|{[\da-fA-F]{4}})|N\{\w+\}|[Pp](?:\{\w+\}|.))(?:[*+?]\??|\{\d+(?:,\d*)?\}\??)?)|\[.*?(?<!\\)\](?:[*+?]\??|\{\d+(?:,\d*)?\}\??)?|(\((?:[^\(\)]++|(?<!\\\\)|(?1))*\)(?:[*+?]\??|\{\d+(?:,\d*)?\}\??)?)/;

    $self->{assembler} //= Regexp::Assemble->new;

    bless $self, $class;
}

sub add {
    my $self = shift;

    $self->{assembler}->add(@_);
}

sub _min_index {
    my $self = shift;

    min($_[0] =~ /\(\?<@{[INDEX_KEYWORD]}>(\d+)\)/g);
}

sub regexp {
    no warnings 'redefine';

    my $self = shift;

    local $Regexp::Assemble::Single_Char = qr//;
    local *Regexp::Assemble::_re_sort = sub($$) {
        my ($a, $b) = @_;

        $self->_min_index($a) <=> $self->_min_index($b)
            || length $b <=> length $a
            || $a cmp $b;
    };

    $self->{assembler}->re;
}

1;


lib/Router/Assemble.pm
package Router::Assemble;

use 5.010001;
use strict;
use warnings;
use utf8;

our $VERSION = '1.00';

use Router::Assemble::Builder;

sub new {
    my $class = shift;
    my $self  = +{@_==1 ? %{$_[0]} : @_};

    my %builder_args;
    for my $k (qw(route_cmp quotemeta_static_path)) {
        $builder_args{$k} = delete $self->{$k} if exists $self->{$k};
    }
    $self->{builder} ||= Router::Assemble::Builder->new(%builder_args);

    bless $self, $class;
}

sub matcher {
    my $self = shift;

    $self->{matcher} ||= $self->{builder}->matcher;
}

sub add {
    my $self = shift;

    delete $self->{matcher};

    $self->{builder}->add(@_);
}

sub match {
    shift->matcher->match(@_);
}

1;
__END__

=encoding utf-8

=head1 NAME

Router::Assemble - Fast, dumpable/restorable and weight adjustable routing engine for web applications

=head1 SYNOPSIS

    use Router::Assemble;

    my $router = Router::Assemble->new();
    $router->add('/', 'dispatch_root');
    $router->add('/entrylist', 'dispatch_entrylist');
    $router->add('/:user', 'dispatch_user');
    $router->add('/:user/{year}', 'dispatch_year');
    $router->add('/:user/{year}/{month:\d+}', 'dispatch_month');
    $router->add('/download/*', 'dispatch_download');

    my $dest = $router->match($env->{PATH_INFO});

=head1 DESCRIPTION

Router::Assemble is yet another routing engine for web applications. This module is inspired by L<Router::Boom> and powered by L<Regexp::Assemble>.

=head2 Fast

This module is pretty fast, almost as fast as L<Router::Boom>.

                         Rate Router::Assemble     Router::Boom
    Router::Assemble 220553/s               --              -0%
    Router::Boom     220554/s               0%               --

=head2 Dump and restore

    my $router = Router::Assemble->new();
    build_routes($router); # add routes for your application

    # Get an instance of Router::Assemble::Matcher.
    # This object holds the minimum information necessary for routing.
    my $matcher = $router->matcher;
    $matcher->match($env->{PATH_INFO});

    # dump and restore
    my $hash = $matcher->to_hash;                     # Dump as a hash object that can be expressed as JSON
    my $m    = Router::Assemble::Matcher->new($hash); # Restore. This is very fast
    $m->match($env->{PATH_INFO});

=head2 Weight adjustable

By default, L<Router::Assemble> routes in order of addition.

    my $router = Router::Assemble->new();
    $router->add('/{category:\w+}', 'cat');
    $router->add('/blog',           'blog_main');

    $router->match('/blog') # cat

By using I<route_cmp>, we can adjust routing order.

    my $router = Router::Assemble->new(
        route_cmp => sub($$){
            my ($a, $b) = @_;
            # route to static route first
            $a->[0] =~ m/[\+\*]/ <=> $b->[0] =~ m/[\+\*]/;
        },
    );
    $router->add('/{category:\w+}', 'cat');
    $router->add('/blog',           'blog_main');

    $router->match('/blog') # blog_main

=head2 Compatibility with L<Router::Boom>

L<Router::Assemble> is compatible with L<Router::Boom> about basic API.

=head1 MEHTODS

=over 4

=item my $router = Router::Assemble->new(%options)

Create a new instance. %options may contains these values.

=over 4

=item route_cmp

A subroutine for comparing routes. this subroutine will be called at first arguments of "sort". The value to be compared is [$path, $context] pair of registered via "add".

=back

=item $router->add($path, $context)

Add a route. $path accepts a string value. See the PATH DEFINITION section for details on how $path value is specified. $context accepts any value.

=item my ($context, $captured) = $router->match($path);

Attempt to find route that matches the given $path and return two values if a route is found. $context is a value of registered via "add". $captured is a hashref that contains placeholders given from $path.

Return empty list if no route is found.

=back

=head1 PATH DEFINITION

=over 4

=item C<{name:\w+}>

Accepts "\w+" in this component. And "name" will be used as a key in the $captured.

    $router->add('/user/{name:\w+}', 'profile');
    $router->match('/user/foo');
    # => ('profile', {name => 'foo'})

You can use "(?:pattern)". (But you can not use "(pattern)")

    $router->add('/user/{name:(?:bar|baz)}', 'special_profile');
    $router->add('/user/{name:\w+}', 'profile');
    $router->match('/user/bar');
    # => ('special_profile', {name => 'bar'})

=item C<:name> and C<{name}>

Same as C<{name:[^/]+}>.

    $router->add('/user/:name', 'profile');
    $router->match('/user/foo');
    # => ('profile', {name => 'foo'})

=item C<*>

Same as C<{*:.+}>.

    $router->add('/archive/*', 'archive');
    $router->match('/archive/path/to/files.zip');
    # => ('archive', {'*' => 'path/to/files.zip'})

=back

=head1 LICENSE

Copyright (C) Six Apart Ltd. E<lt>sixapart@cpan.orgE<gt>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Taku AMANO E<lt>usualoma@cpan.orgE<gt>

=head1 SEE ALSO

L<Router::Boom> Pretty fast routing engine for web applications.

L<Regexp::Assemble> Assemble multiple Regular Expressions into a single RE.

=cut
